File: refsol/maximum-product-subarray.py
def maxProduct(nums):
    result = nums[0]
    cur_max = nums[0]
    cur_min = nums[0]
    for n in nums[1:]:
        candidates = (n, cur_max * n, cur_min * n)
        cur_max = max(candidates)
        cur_min = min(candidates)
        result = max(result, cur_max)
    return result

File: refsol/maximum-product-subarray.js
export function maxProduct(nums) {
  // Independent approach: scan products from both directions.
  const n = nums.length;
  let best = nums[0];
  let prefix = 0;
  let suffix = 0;
  for (let i = 0; i < n; i++) {
    prefix = (prefix === 0 ? 1 : prefix) * nums[i];
    suffix = (suffix === 0 ? 1 : suffix) * nums[n - 1 - i];
    best = Math.max(best, prefix, suffix);
  }
  return best;
}
